### PR TITLE
Continue refactoring of RSA crate internals

### PR DESCRIPTION
**Pull request by lumag**
Further refactoring of the RSA crate. Clean up internal interfaces, split software RSA implementation from the wrapping code.

### CODE REVIEW

**Comment by lumag** [comment]
cc @dignifiedquire @tarcieri @str4d

File: src/algorithms/rsa.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+use num_traits::{One, Signed, Zero};
+use rand_core::CryptoRngCore;
+use zeroize::Zeroize;
+
+use crate::errors::{Error, Result};
+use crate::keytraits::{PrivateKeyParts, PublicKeyParts};
+
+/// Raw RSA encryption of m with the public key. No padding is performed.
+#[inline]
+pub fn rsa_encrypt<K: PublicKeyParts>(key: &K, m: &BigUint) -> Result<BigUint> {
```
Comment (tarcieri) [inline src/algorithms/rsa.rs:15]: This feels like a hazmat thing, and should probably carry a warning about the dangers of misused unpadded RSA

Comment (lumag) [reply src/algorithms/rsa.rs:15]: We can not put it under hazmat as it is used internally. Please note that that all algorithms modules are `pub(crate)` at best. But yes, the warning is probably a good idea.

File: src/lib.rs [hunk 2]
```diff
@@ -222,8 +222,9 @@ pub use num_bigint::BigUint;
 pub use rand_core;
 pub use signature;
 
-pub mod algorithms;
+mod algorithms;
```
Comment (tarcieri) [inline src/lib.rs:225]: Aah, ok. Maybe make the methods within explicitly `pub(crate)` for clarity?

File: src/algorithms/rsa.rs (hunk 1 above)
Comment (tarcieri) [reply src/algorithms/rsa.rs:15]: Now I see: https://github.com/RustCrypto/RSA/pull/304#pullrequestreview-1392921745

File: src/lib.rs (hunk 2 above)
Comment (lumag) [reply src/lib.rs:225]: Yep. Sounds like a good idea. I'll change that and add a warning too.

Comment (lumag) [reply src/lib.rs:225]: done

**Comment by tarcieri** [comment]
@lumag need to remove `expose-internals` here: https://github.com/RustCrypto/RSA/blob/master/.github/workflows/ci.yml#L70

**Comment by lumag** [comment]
done

**Review by tarcieri — APPROVED** [review]

**Pull request merged**